fedimint-cli/src/lib.rs

```diff
@@ -506,6 +506,7 @@ impl FedimintCli {
                 let _ = writeln!(std::io::stdout(), "{output}");
             }
             Err(err) => {
+                let _ = writeln!(std::io::stdout(), "{err}");
```

Comment:
stdout should output only json, so we want to wrap the error in some kind of:

```json
{
  error: "Things failed"
}
```

and in the future evolve to:

```
```json
{
  error: "Things failed"
  code: 134,
  hint: "Try again on Monday"
}
```

Reply:
Let me know if I'm misinterpreting, but `err` is a `CliError` (https://github.com/fedimint/fedimint/blob/cd9b5a2/fedimint-cli/src/lib.rs#L124-L133) so this should output only json unless I'm misinterpreting.

```rust
/// Cli error
#[derive(Serialize, Error)]
#[serde(tag = "error", rename_all(serialize = "snake_case"))]
struct CliError {
    kind: CliErrorKind,
    message: String,
    #[serde(skip_serializing)]
    #[source]
    raw_error: Option<anyhow::Error>,
}
```

Reply:
Is that how it is being printed already? I assumed `{err}` will format it as a human readable string instread of json. I haven't checked. If you have a good way to trigger it in devimint env, please let me know so I can confirm when I get to it.

Reply:
> Is that how it is being printed already?

Yep, `CliError` implements `fmt::Display` which converts to json (https://github.com/fedimint/fedimint/blob/cd9b5a2/fedimint-cli/src/lib.rs#L219-L227)

```rust
impl fmt::Display for CliError {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        let mut json = serde_json::to_value(self).unwrap();
        if let Some(err) = &self.raw_error {
            json["raw_error"] = json!(*err.to_string())
        }
        write!(f, "{}", serde_json::to_string_pretty(&json).unwrap())
    }
}
```

> If you have a good way to trigger it in devimint env, please let me know so I can confirm when I get to it.

Thanks for asking! This uncovered how awkward this PR makes the output when interacting with the cli. You can repro in devimint env using

```bash
> just mprocs
> fedimint-cli spend 10000000
(this will succeed)
...
> fedimint-cli spend 10000000
2024-03-15T03:10:26.559703Z  WARN fedimint_cli::client: The client will try to double-spend these notes after the duration specified by the --timeout option to recover any unclaimed e-cash.
{
  "error": "CliError",
  "kind": "general_failure",
  "message": "failure",
  "raw_error": "Insufficient balance: requested 10000000 msat but only 0 msat available"
}
{
  "error": "CliError",
  "kind": "general_failure",
  "message": "failure",
  "raw_error": "Insufficient balance: requested 10000000 msat but only 0 msat available"
}
```

The issue I was hoping to resolve is parsing the json error in devimint tests, which we can't do cleanly using stderr since it contains the logs.

You can verify by redirecting stdout to /dev/null

```bash
> fedimint-cli spend 10000000 > /dev/null
2024-03-15T03:14:22.166187Z  WARN fedimint_cli::client: The client will try to double-spend these notes after the duration specified by the --timeout option to recover any unclaimed e-cash.
{
  "error": "CliError",
  "kind": "general_failure",
  "message": "failure",
  "raw_error": "Insufficient balance: requested 10000000 msat but only 0 msat available"
}
```

I think a different approach that may work better is to set `RUST_LOG=off` specifically when calling `expect_err_json` in devimint tests, then parsing the error from stderr.

Rubber ducking at this point, I'll ping for another review when this is cleaner 😄

Reply:
IO we should just log the error (only the `raw_error`  part) as `error!(...) on stderr, and not the whole json anymore, as it's now redundant. @bradleystachurski 

So it will look like:

```
2024-03-15T05:13:18.774024Z  INFO task: Task finished task=state machine executor
2024-03-15T05:13:18.774024Z ERROR fedimint-cli: Insufficient balance: requested 10000000 msat but only 0 msat available
{
  "error": "CliError",
  "kind": "general_failure",
  "message": "failure",
  "raw_error": "Insufficient balance: requested 10000000 msat but only 0 msat available"
}
```

Reply:
BTW. This error json structure doesn't make sense to me.

The human readable error should go into `error`.

The `kind` seems suspiciously useless, and `message` worthless.

I don't know if we copied it from some existing project that had a better use for these, or something, but if this is to be parsed now by our tests, it's the last chance to improve it for free.

Reply:
@elsirion If you're around.

Reply:
I could imagine `kind` being used to distinguish between errors originating in modules vs. `fedimint-cli`? But I think the way it's used rn is redundant. Probably best to start over and cut the error struct down to the minimum: `{"error": "Some serialized error …"}`.

Reply:
True. Adding a field is backward-compat, but removing it always potentially breaks someone.

Reply:
Updated so we have a minimal error struct printed on stdout and the error message logged as an error on stderr:

```bash
> just mprocs
> fedimint-cli spend 10000000
(this will succeed)
...
> fedimint-cli spend 10000000
2024-03-16T20:59:08.258889Z  WARN fedimint_cli::client: The client will try to double-spend these notes after the duration specified by the --timeout option to recover any unclaimed e-cash.
2024-03-16T20:59:08.275700Z ERROR fedimint_cli: Insufficient balance: requested 10000000 msat but only 0 msat available
{
  "error": "Insufficient balance: requested 10000000 msat but only 0 msat available"
}
```